### DISCUSSION TITLE
handlebars.js - Week Year Number

### DISCUSSION THREAD

**Discussion by insomniacj93** [Budibase/budibase#6368] [Help] [answered]
Hello there, dear friends!

I'm looking for help with the handlebars, I'm trying to generate a unique ID for every entry. I'm retrieving data from the user input and one thing i really wanted to do is get the current week number (the number of the current week that the ID is being generated "1-52").

This is what my handlebars.js looks like atm:

{{ Form.Fields.product_id }}{{ date now "YYMMDD" }}{{ random 10 99 }}{{ "PT" }} 

=Output:      "PRODUCT22061799PT"       (current date = 22-06-17)

I like how it looks so far, but my intention is to ditch the "MM" and the "DD" and replace them with just the Week Year. I found some documentation online about the handlebars.js and they reference Week Year as "GG" but such acronym isnt recognized in the budibase handlebars compiler.

ExpectedResult:         "PRODUCT225299PT"        (22 = Year / 52 = Week Year)

myref: https://docs.celigo.com/hc/en-us/articles/360045564992-Handlebar-expressions-for-date-and-time-format-codes

Any help on how to resolve this issue?

Very appreciated!!

**Comment by melohagan** [comment, marked as answer]
Hi @insomniacj93 

This does look like a bug which I've raised here: https://github.com/Budibase/budibase/issues/6369

In the meantime, I found this JavaScript code snippet which may be of use instead of a handlebars expression:
https://www.w3resource.com/javascript-exercises/javascript-date-exercise-24.php

**Reply by insomniacj93** [reply]
Hello, @melohagan 

Thank you for your help, I've tried to add the javascript but my placeholder doesnt return any value, I've edited it with return to get the console.log but it didnt work. Also is there any solution to the bug, or is there any plan to solve it?

Honestly I just wish i could get the same result "PRODUCT225299PT". With either handlebars.js or pure javascript.

Best Regards.

**Reply by melohagan** [reply]
This worked for me:

```
function ISO8601_week_no(dt) 
  {
     var tdt = new Date(dt.valueOf());
     var dayn = (dt.getDay() + 6) % 7;
     tdt.setDate(tdt.getDate() - dayn + 3);
     var firstThursday = tdt.valueOf();
     tdt.setMonth(0, 1);
     if (tdt.getDay() !== 4) 
       {
      tdt.setMonth(0, 1 + ((4 - tdt.getDay()) + 7) % 7);
        }
     return 1 + Math.ceil((firstThursday - tdt) / 604800000);
        }

function randomIntFromInterval(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min)
}

return ISO8601_week_no(new Date()) + "" + randomIntFromInterval(10, 99) + "PT"
```

It just needs the form field product id at the start of the returned string.
Also the returned week number is 25, but that makes sense because it's June. I'm not sure why you would expect 52?

**Reply by insomniacj93** [reply]
@melohagan 

oh, no.

When i said 52 it was a random week number, 25 is what we should be expecting, yes!

Thank you very much, I'll test it out in the forms.

**Comment by insomniacj93** [comment]
@melohagan 

It works perfect!!

Just one more thing.. Perhaps is it possible we set the random number as a string so that all the 0's in the left also count?
The main reason why I randomized from 10 to 99 was because a value like "03" wouldnt print out as "03" but just "3".
But i really need to keep the length of the whole string.

Thanks in advance for all your time!!

**Reply by melohagan** [reply]
You may be able to use the padStart method: https://stackoverflow.com/questions/10073699/pad-a-number-with-leading-zeros-in-javascript

There's a few other options if not.

**Comment by insomniacj93** [comment]
@melohagan 
Hi,
The ID is looking great!
I'm currently looking for a way to do a while loop (increment +1) instead of the random number generator (randomIntFromInterval).
This means everytime we execute the code it will increment+1 keeping in mind the number of the week (we keep incrementing the value as long as we are in the same week).

[CODE]

...

function randomIntFromInterval(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min)
}

return ISO8601_week_no(new Date()) + $("Form.Fields.product_id")  + randomIntFromInterval(10, 99) + "PT"



Thank you!
